mlprimitives/candidates/dsp.py

```diff
@@ -0,0 +1,211 @@
+import numpy as np
+
+
+def next_power_of_2(x):
+    """Finds the next power of 2 value
+
+    Args:
+        x: Input value
+
+    Returns:
+        power_of_2: Next power of 2 value
+
+    """
+
+    power_of_2 = 1 if x == 0 else 2 ** np.ceil(np.log2(x))
+    return power_of_2
+
+
+class SpectralMask:
+    """Anomalies detection in satellite telemetry data using a spectral mask"""
+
+    def __init__(self, method='std_dev', gain=1, window_length=128, beta=4):
+        """Inits SpectralMask Class with default values for method, gain, window_length, and
+             beta
+
+        Args:
+            method: Method used to calculate the Spectral Mask: 'std_dev' uses the standard
+              deviation of each frequency component and 'min_max' uses the minimum and maximum
+              values
+            gain: Multiplication factor used in the comparison between the spectral mask defined
+              using the training data and the Fourier transform of the telemetry data vector
+            window_length: Length of the sliding window in number of samples
+            beta: Beta value for Kaiser window design
+
+        """
+
+        self.gain = gain
+        self.beta = beta
+        self.window_length = window_length
+        self.window = None
+        self.method = method
+        self.window_length = next_power_of_2(window_length)
+
+    def window_design(self, window_length=None, beta=None):
+        """Kaiser window design
+
+        Args:
+            window_length: Length of the window in number of samples
+            beta: Beta value for Kaiser window design
+
+        Returns:
+            window: Window designed using the beta and length provided as inputs
+
+        """
+
+        if window_length is None:
+            window_length = self.window_length
+        else:
+            self.window_length = window_length
+
+        if beta is None:
+            beta = self.beta
+        else:
+            self.beta = beta
+
+        self.window = np.kaiser(window_length, beta)
+
+        return self.window
+
+    def fit(self, X, method=None, gain=None, window_length=None, beta=None):

```

Comment:
Since most of the arguments will only be passed to the constructor, as class-level hyperparameters, I'd recommend removing them from here, which will allow simplifying a lot the code (all the if/else checks will go away)

Reply:
I usually test these methods outside mlblocks, using my own utils, calling different methods with different parameters and comparing results several times after the object is created. I'd like to keep these parameters while I'm still developing, otherwise, the working version and the one uploaded are going to be too different. I can remove them in the final version if you think it's better. Is that ok?

Reply:
I don't really see much difference when it comes to testing.

Simply this:

```
sm = SpectralMask(method, gain, window_length, beta)
sm.fit(X)
```

becomes this, which is pretty much the same:

```
sm = SpectralMask()
sm.fit(X, method, gain, window_length, beta)
```

And when it comes to produce, you can even do:

```
sm = SpectralMask()
sm.mask_top = mask_top
sm.mask_bottom = mask_bottom
sm.window = window
sm.produce(X)
```

But, sure, if it helps your testing while the version is not final and the primitive can be used anyway, I see no harm in it, as far as it works when used as a primitive and the end result is clean and easy to understand.

Reply:
my understanding from previous conversations with the team was that the only way to pass hyperparameters to the object is via the constructor of the class. Are you telling me that the fit and produce methods can receive these parameters directly?

Reply:
No, I meant the opposite: When using the primitive, the hyperparameters need to be passed to the constructor, so this is OK right now.
However, you said that you want to be able to pass the values to the methods directly for testing. What I was saying is that I see no need on that: Whichever values you want to pass to the methods can be also passed via the constructor during testing. And if what you need are the attributes, you can also set them on the fly for testing.